minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java

```diff
@@ -0,0 +1,426 @@
+  private Closer mCloser;
+  private List<Master> mMasters;
+  private List<Worker> mWorkers;
+  /** Addresses of all masters. Should have the same size as {@link #mMasters}. */
+  private List<MasterNetAddress> mMasterAddresses;
+  private State mState;
+  private TestingServer mCuratorServer;
+
+  private MultiProcessCluster(Map<PropertyKey, String> properties, int numMasters, int numWorkers,
+      String clusterName) {
+    mProperties = properties;
+    mNumMasters = numMasters;
+    mNumWorkers = numWorkers;
+    // Add a unique number so that different runs of the same test use different cluster names.
+    mClusterName = clusterName + ThreadLocalRandom.current().nextLong();
+    mMasters = new ArrayList<>();
+    mWorkers = new ArrayList<>();
+    mCloser = Closer.create();
+    mState = State.NOT_STARTED;
+  }
+
+  /**
+   * Starts the cluster, launching all server processes.
+   */
+  public synchronized void start() throws Exception {
+    Preconditions.checkState(mState != State.STARTED, "Cannot start while already started");
+    Preconditions.checkState(mState != State.DESTROYED, "Cannot start a destroyed cluster");
+    mState = State.STARTED;
+
+    mMasterAddresses = generateMasterAddresses(mNumMasters);
+    if (zkEnabled()) {
+      mCuratorServer = mCloser
+          .register(new TestingServer(-1, AlluxioTestDirectory.createTemporaryDirectory("zk")));
+      mProperties.put(PropertyKey.ZOOKEEPER_ADDRESS, mCuratorServer.getConnectString());
+    } else {
+      MasterNetAddress masterAddress = mMasterAddresses.get(0);
+      mProperties.put(PropertyKey.MASTER_HOSTNAME, masterAddress.getHostname());
+      mProperties.put(PropertyKey.MASTER_RPC_PORT, Integer.toString(masterAddress.getRpcPort()));
+      mProperties.put(PropertyKey.MASTER_WEB_PORT, Integer.toString(masterAddress.getWebPort()));
+    }
+
+    mWorkDir = AlluxioTestDirectory.createTemporaryDirectory(mClusterName);
+    for (Entry<PropertyKey, String> entry : ConfigurationTestUtils.testConfigurationDefaults(
+        NetworkAddressUtils.getLocalHostName(), mWorkDir.getAbsolutePath()).entrySet()) {
+      // Don't overwrite explicitly set properties.
+      if (!mProperties.containsKey(entry.getKey())) {
+        mProperties.put(entry.getKey(), entry.getValue());
+      }
+    }
+
+    new File(Configuration.get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS)).mkdirs();
+    formatJournal();
+    writeConf();
+
+    // Start servers
+    LOG.info("Starting alluxio cluster {} with base directory {}", mClusterName,
+        mWorkDir.getAbsolutePath());
+    for (int i = 0; i < mNumMasters; i++) {
+      startMaster(i);
+    }
+    for (int i = 0; i < mNumWorkers; i++) {
+      startWorker(i);
+    }
+  }
+
+  /**
+   * Kills the primary master.
+   *
+   * If no master is currently primary, this method blocks until a primary has been elected, then
+   * kills it.
+   *
+   * @return the ID of the killed master
+   */
+  public synchronized int killPrimaryMaster() {
```

Comment:
nit: `killPrimaryMasterOrWait`?

Reply:
changed name to `waitForAndKillPrimaryMater`